Add StimulusJS

oliverguenther opened the pull request:
- [x] Adds stimulus to the bundle
- [x] Adds an application controller to allow lazy loading controller chunks in the same fashion as dynamic scripts
- [x] Rewrites the `admin_users` dynamic script to use stimulus controllers 
- [x] Adds `strictBindCallApply` to the tsconfig to allow us to use `bind(this)` without linter errors

https://community.openproject.org/wp/48130


**Scripts to convert**
- [x] administration_settings.js [OG]
- [x] admin_users.js [OG]
- [x] backlogs.js [OG]
- [ ] custom_fields.js [quite big, might make sense in its own PR]
- [x] forums.js [OG]
- [x] global_roles.ts [JU]
- [x] meeting.js [JU]
- [ ] members_form.js [JU]
- [x] new_user.js
- [x] print_handler.js
- [x] project_folder_mode_selection.ts
- [x] project_form_listener.js
- [ ] project.js [JU - but quite big, maybe separate PR]
- [x] repository_navigation.js [OG]
- [x] repository_settings.js [OG]
- [x] two_factor_authentication.ts [OG]

oliverguenther committed:
admin users controller

Add helper to determine when async controllers are loaded

remove unnecessary stimulus connected spec helper

oliverguenther committed:
Add generalized disable-when-checked controller

oliverguenther committed:
Enable strictBindCallApply

This allows us to use .bind(this) without errors

ulferts commented on frontend/src/stimulus/controllers/op-application.controller.ts:
```diff
@@ -0,0 +1,32 @@
+
+  public initialize():void {
+    this.bootstrapDynamicControllers();
+  }
```
Wouldn't it be more resilient if a `dynamicTargetConnected` were to be used? The `initialize` callback is only called once when the controller is initialized. But with this controller residing at the top level of the DOM structure, the DOM governed by it might change later on by server responses. Then it would be good for the controller to also pick up the controllers to be loaded for the added DOM. 

For now, this definitely not a problem but it would be interesting to see if the approach chosen here will also work with dynamically added DOM.

oliverguenther replied:
Indeed that would be better, I changed it to that approach

ulferts commented on spec/support/stimulus/stimulus_helpers.rb:
```diff
@@ -0,0 +1,37 @@
+  def expect_stimulus_controller_connected(name, connected: true)
+    expect(page).to have_selector("[data-controller='#{name}']") { |el| el['data-connected'] == connected.to_s }
+  end
+end
```
What is the reason for testing for the stimulus binding? A feature spec should focus on the effects and not on the implementation IMO.

oliverguenther replied:
I used it in a previous variant where the auth source would connect as a target, and then the selection might happen before the controller is bound. This is no longer the case so I'll remove it

ulferts commented on frontend/src/stimulus/controllers/disable-when-checked.controller.ts:
```diff
@@ -0,0 +1,30 @@
+    target.removeEventListener('change', this.handleChange.bind(this));
+  }
+
+  private handleChange(evt:InputEvent):void {
```
Matter of preference I guess but describing what is done in the method is preferable to me (e.g. `toggleDisabled` or `disableWhenChecked`)

ulferts reviewed (commented):
I admittedly have to wrap my head around this style of writing interactions in JS again. Expecting to be able to bind the disabled state of an input to the controller, I looked into the documentation and was surprised that this is actually not what Stimulus offers. 

Spreading controller definitions and targets over multiple views/partials is also something I will have to get used to.

On the plus side, the implementation is straight forward so once I freshened up on Stimulus I was able to understand it quickly.

oliverguenther committed:
Add stimulus.js

Use dynamicTargetLoaded to register controllers

ulferts reviewed (approved):
👍

oliverguenther committed:
Remove administration_settings script

It is no longer valid since available languages are on a different page
than user default language

oliverguenther committed:
Move backlogs dynamic scripts to a lazyloaded stimulus controller

ulferts committed:
employ stimulus controller on role admin page

ulferts committed:
employ stimulus controller on journal history pages

oliverguenther committed:
Replace quoting messages scripts with stimulus

oliverguenther committed:
Replace 2FA script with stimulus controller

ulferts committed:
Replace meeting script with stimulus controller

ulferts commented on frontend/src/stimulus/controllers/dynamic/meeting-content.controller.ts:
```diff
@@ -0,0 +1,82 @@
+  ];
+
+  initialize() {
+    // This should have been achieved by the `data-loader-edit-state-value` attribute, but for unknown reasons
+    // this does not apply the value to the variable. Doing it manually for now.
+    const rawInitialValue = this.element.getAttribute('data-loader-edit-state-value');
+
+    if (rawInitialValue === 'true') {
+      this.editStateValue = true;
+    } else if (rawInitialValue === 'false') {
+      this.editStateValue = false;
+    }
```
I added this workaround for what should have worked out of the box. I actually applied the `data-loader...` attribute successfully for the `admin-roles.controller` and cannot spot any difference. Do you have an idea, @oliverguenther

oliverguenther replied:
The values are not called `data-loader`, but `data-[controllername]-[valuename]`. so this would be `data-meeting-content-state-value="..."`.  That's why it wasn't picked up

cf., https://stimulus.hotwired.dev/reference/values#properties-and-attributes

ulferts replied:
Ah, I was fooled by the example on that page (it is called loader which sounded like what I was trying to achieve, load data). 
Thanks @oliverguenther

ulferts committed:
properly use stimulus data loading

ulferts committed:
adapt wiki specs

ulferts commented:
The spec failing currently is ultimately caused in `dev`. I fixed it there and expect the spec to turn green again.

ulferts committed:
apply admin-users.controller to new user form

ulferts committed:
adapt wiki specs

oliverguenther committed:
Replace repository-settings

oliverguenther committed:
Remove unused project_form_listener

oliverguenther committed:
Replace print_handler with print controller

oliverguenther committed:
Add option to provide controller for entire page with helper

oliverguenther committed:
Replace repository_navigation with stimulus

oliverguenther committed:
Replace project_folder_mode_selection

oliverguenther committed:
Replace with content_controller and allow passing values
